core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -0,0 +1,477 @@
+
+	/**
+	 * Adds RDF data from an {@link InputStream} to the transaction.
+	 * 
+	 * @param inputStream
+	 * @param baseURI
+	 * @param format
+	 * @param contexts
+	 * @throws ExecutionException
+	 * @throws InterruptedException
+	 */
+	void add(InputStream inputStream, String baseURI, RDFFormat format, boolean preserveBNodes,
+			Resource... contexts)
+		throws InterruptedException, ExecutionException
+	{
+		Future<Boolean> result = executor.submit(() -> {
+			try {
+				if (preserveBNodes) {
+					// create a reconfigured parser + inserter instead of relying on standard
+					// repositoryconn add method.
+					RDFParser parser = Rio.createParser(format);
+					parser.getParserConfig().set(BasicParserSettings.PRESERVE_BNODE_IDS, true);
+					RDFInserter inserter = new RDFInserter(txnConnection);
+					inserter.setPreserveBNodeIDs(true);
+					if (contexts.length > 0) {
+						inserter.enforceContext(contexts);
+					}
+					parser.setRDFHandler(inserter);
+					parser.parse(inputStream, baseURI);
+				}
+				else {
+					txnConnection.add(inputStream, baseURI, format, contexts);
+				}
+				return true;
+			}
+			catch (IOException e) {
+				throw new RuntimeException(e);
+			}
+		});
+
+		futures.add(result);
+		result.get();
+	}
+
+	/**
+	 * @param contentType
+	 * @param inputStream
+	 * @param baseURI
+	 * @throws ExecutionException
+	 * @throws InterruptedException
+	 */
+	void delete(RDFFormat contentType, InputStream inputStream, String baseURI)
+		throws InterruptedException, ExecutionException
+	{
+		Future<Boolean> exception = executor.submit(() -> {
+			RDFParser parser = Rio.createParser(contentType, txnConnection.getValueFactory());
+
+			parser.setRDFHandler(new WildcardRDFRemover(txnConnection));
+			parser.getParserConfig().set(BasicParserSettings.PRESERVE_BNODE_IDS, true);
+			try {
+				parser.parse(inputStream, baseURI);
+				return true;
+			}
+			catch (IOException e) {
+				throw new RuntimeException(e);
+			}
+		});
+
+		futures.add(exception);
+
+		exception.get();
+	}
+
+	/**
+	 * @param queryLn
+	 * @param sparqlUpdateString
+	 * @param baseURI
+	 * @param includeInferred
+	 * @param dataset
+	 * @param bindings
+	 * @throws ExecutionException
+	 * @throws InterruptedException
+	 */
+	void executeUpdate(QueryLanguage queryLn, String sparqlUpdateString, String baseURI,
+			boolean includeInferred, Dataset dataset, Map<String, Value> bindings)
+		throws InterruptedException, ExecutionException
+	{
+		Future<Boolean> result = executor.submit(() -> {
+			Update update = txnConnection.prepareUpdate(queryLn, sparqlUpdateString);
+			update.setIncludeInferred(includeInferred);
+			if (dataset != null) {
+				update.setDataset(dataset);
+			}
+			for (String bindingName : bindings.keySet()) {
+				update.setBinding(bindingName, bindings.get(bindingName));
+			}
+
+			update.execute();
+			return true;
+		});
+
+		futures.add(result);
+		result.get();
+	}
+
+	boolean hasActiveOperations() {
+		for (Future future : futures) {
```

Comment:
Now that I look at it - this may need to be synchronized on `futures` - and all methods that create/add a new `Future` should also synchronize on it.